atomic_reactor/plugins/post_group_manifests.py

```diff
@@ -47,144 +59,298 @@ def __init__(self, tasker, workflow, registries, group=True, goarch=None):
-                raise
-            self.log.info("Manifest list submitted for %s", registry)
-
-    def get_grouped_manifests(self):
-        grouped_manifests = []
-        for registry, registry_conf in self.registries.items():
-            if registry_conf.get('version') == 'v1':
-                continue
+    def get_manifest(self, session, repository, ref):
+        """
+        Downloads a manifest from a registry. ref can be a digest, or a tag.
+        """
+        self.log.debug("%s: Retrieving manifest for %s:%s", session.registry, repository, ref)
+
+        headers = {
+            'Accept': ', '.join((
+                MEDIA_TYPE_MANIFEST_V2,
+                MEDIA_TYPE_LIST_V2,
+                MEDIA_TYPE_OCI,
+                MEDIA_TYPE_OCI_INDEX
+            ))
+        }
+
+        url = '/v2/{}/manifests/{}'.format(repository, ref)
+        response = session.get(url, headers=headers)
+        response.raise_for_status()
+        return (response.content,
+                response.headers['Docker-Content-Digest'],
+                response.headers['Content-Type'],
+                int(response.headers['Content-Length']))
+
+    def link_blob_into_repository(self, session, digest, source_repo, target_repo):
+        """
+        Links ("mounts" in Docker Registry terminology) a blob from one repository in a
+        registry into another repository in the same registry.
+        """
+        self.log.debug("%s: Linking blob %s from %s to %s",
+                       session.registry, digest, source_repo, target_repo)
+
+        # Check that it exists in the source repository
+        url = "/v2/{}/blobs/{}".format(source_repo, digest)
+        result = session.head(url)
+        if result.status_code == requests.codes.NOT_FOUND:
+            self.log.debug("%s: blob %s, not present in %s, skipping",
+                           session.registry, digest, source_repo)
+            # Assume we don't need to copy it - maybe it's a foreign layer
+            return
+        result.raise_for_status()
+
+        url = "/v2/{}/blobs/uploads/?mount={}&from={}".format(target_repo, digest, source_repo)
+        result = session.post(url, data='')
+        result.raise_for_status()
+
+        if result.status_code != requests.codes.CREATED:
+            # A 202-Accepted would mean that the source blob didn't exist and
+            # we're starting an upload - but we've checked that above
+            raise RuntimeError("Blob mount had unexpected status {}".format(result.status_code))
+
+    def link_manifest_references_into_repository(self, session, manifest, media_type,
+                                                 source_repo, target_repo):
+        """
+        Links all the blobs referenced by the manifest from source_repo into target_repo. If
+        manifest_digest is passed, also links the manifest itself. (As a *blob* not as
+        a manifest... to use it directly as a manifest, it needs to be uploaded with a tag.)
```

Comment:
Not sure I understand the manifest_digest remark.

Reply:
Sorry, a left-over from older versions of the code. Removed in the latest push.